src/avram/operation.cr

```diff
@@ -1,23 +1,94 @@
 require "./validations"
+require "./callbacks"
 require "./define_attribute"
 require "./save_operation_errors"
 require "./param_key_override"
+require "./needy_initializer"
 
-class Avram::VirtualForm
-  macro inherited
-    {% raise "Avram::VirtualForm has been renamed to Avram::Operation. Please inherit from Avram::Operation." %}
-  end
-end
-
-class Avram::Operation
+abstract class Avram::Operation
+  include Avram::NeedyInitializer
   include Avram::DefineAttribute
   include Avram::Validations
   include Avram::SaveOperationErrors
   include Avram::ParamKeyOverride
+  include Avram::Callbacks
 
   @params : Avram::Paramable
   getter params
 
+  # Yields the instance of the operation, and the return value from
+  # the `run` instance method.
+  #
+  # ```
+  # MyOperation.run do |operation, value|
+  #   # operation is complete
+  # end
+  # ```
+  def self.run(*args, **named_args)
+    params = Avram::Params.new
+    run(params, *args, **named_args) do |operation, value|
+      yield operation, value
+    end
+  end
+
+  # Returns the value from the `run` instance method.
+  # or raise `Avram::FailedOperation` if the operation fails.
+  #
+  # ```
+  # value = MyOperation.run!
+  # ```
+  def self.run!(*args, **named_args)
+    params = Avram::Params.new
+    run!(params, *args, **named_args)
+  end
+
+  # Yields the instance of the operation, and the return value from
+  # the `run` instance method.
+  #
+  # ```
+  # MyOperation.run(params) do |operation, value|
+  #   # operation is complete
+  # end
+  # ```
+  def self.run(params : Avram::Paramable, *args, **named_args)
+    operation = self.new(params, *args, **named_args)
+    operation.before_run
+    value = operation.run
+    if operation.valid?
+      operation.after_run(value)
+    else
+      value = nil
+    end
+    yield operation, value
+  end
+
+  # Returns the value from the `run` instance method.
+  # or raise `Avram::FailedOperation` if the operation fails.
+  #
+  # ```
+  # value = MyOperation.run!(params)
+  # ```
+  def self.run!(params : Avram::Paramable, *args, **named_args)
+    operation = self.new(params, *args, **named_args)
+    operation.before_run
+    value = operation.run
+    if operation.valid?
+      operation.after_run(value)
+    else
+      value = nil
+    end
+    raise Avram::FailedOperation.new("The operation failed to return a value") unless value
+    operation
+  end
```

Comment:
We can delegate to `.run`
```suggestion
  def self.run!(params : Avram::Paramable, *args, **named_args)
    run(params, *args, **named_args) do |operation, value|
      raise Avram::FailedOperation.new("The operation failed to return a value") unless value
      operation
    end
  end
```

By the way, should this be returning the operation or the value?

Reply:
My original intention was to return the operation itself in case you needed to use that object for whatever. You'd still have access to the value. But now that I think about it, doing `SaveOperation.create!` returns the model (the value), so I guess this may as well follow suit. If you need access to the operation, you just have to use the block form. I think that's fine, right?